widgetize: Convert module to typescript

joydeep2001 opened the pull request:
widgetize: Convert module to typescript

joydeep2001 commented:
@zulipbot add "area: typescript migration"

LaPulgaaa commented:
Thanks a lot for working on this @joydeep2001 . I have added few comments which might be helpful before core maintainers reviews this . Also I could not understand your motivation behind having last three commits as separate commits . Sorry if I am missing something??

joydeep2001 commented:
I thought it would be good to make smaller commits than converting the whole file in one commit. But as the last three commits doing similar task I can consider squashing them if required. Thanks for pointing that out. 

And Sorry I can't find the comments you added.

LaPulgaaa commented:
> I thought it would be good to make smaller commits than converting the whole file in one commit. But as the last three commits doing similar task I can consider squashing them if required. Thanks for pointing that out.
>
Thanks !. Yes,I don't think those are helpful separations for anyone as they revolve around a similar idea. 
> And Sorry I can't find the comments you added.
>
Apologies....Should be visible now, I forgot to submit my review .

joydeep2001 commented on web/src/widgetize.ts:
```diff
@@ -107,16 +106,16 @@ export function activate(in_opts: WidgetOptions): void {
     for (const [idx, $widget_elem] of widget_contents) {
-        if (message_lists.current.get(idx) !== undefined) {
-            const $row = message_lists.current.get_row(idx);
+        if (message_lists.current!.get(idx) !== undefined) {
```
I agree with  the `assert` idea. But do you think that `const message_list = message_list.current;` will have any extra benefit ? Because `message_list.current` is not an function call or something like that which will require extra computation. Please correct me if I'm getting it wrong. thanks

LaPulgaaa replied:
`message_list.current` is not a call ...Agreed. I suggested this for readability purpose. But I don't think it won't make a large difference ,just `asserting` this upfront was important.

joydeep2001 commented:
@LaPulgaaa Thanks for the your valuable reviews. I've addressed all your comments in the recent push also restructured the commit history as per your suggestion. It would be great if you give it a looks once more.

LaPulgaaa commented on web/src/widgetize.ts:
```diff
@@ -1,17 +1,45 @@
+              heading: string;
+              choices: {type: string; reply: string; long_name: string; short_name: string}[];
+          }
+        | undefined;
```
Should `undefined` be replaced with `null` ?? .

joydeep2001 replied:
I've some confusion with the type of this `extra_data`. I'll check and get back to this. Thanks!

LaPulgaaa commented on web/src/widgetize.ts:
```diff
@@ -1,17 +1,45 @@
+    activate: (data: {
+        $elem: JQuery;
+        callback: (data: string) => void;
+        message: WidgetOptions["message"];
```
Is there a difference b/w `WidgetOptions["message"]` and mere `Message` type. I think Message would be a bit more clear.

joydeep2001 replied:
yah correct! Idk why I had written it like this. will fix.

LaPulgaaa commented on web/src/widgetize.ts:
```diff
@@ -42,23 +70,23 @@ export function activate(in_opts) {
     }
 
-    let $widget_elem = widget_contents.get(message.id);
+    let $widget_elem: JQuery | undefined = widget_contents.get(message.id);
```
I think TS will automatically infer this given that `widget_contents` has been typed.

joydeep2001 replied:
yah...this seems redundant here..

timabbott commented:
@N-Shar-ma perhaps you can review this; feel free to review in the form of a revised PR that you're happy with; that's sometimes easier for TypeScript conversions.

timabbott commented:
@LaPulgaaa maybe you can help review this given @N-Shar-ma has been busy recently.

LaPulgaaa commented on web/src/widgetize.ts:
```diff
@@ -2,15 +2,42 @@ import $ from "jquery";
+type WidgetValue = Record<string, unknown> & {
+    activate: (data: {
+        $elem: JQuery;
+        callback: (data: string) => void;
```
Is the type a `data` parameter always a `string`? It might me for `todo_widget` but for `poll`  `data` looks something like 
```
data: NewOptionOutboundData | QuestionOutboundData | VoteOutboundData | undefined
```
Will be nice if you can cross check and complete this type.

LaPulgaaa commented on web/src/widgetize.ts:
```diff
@@ -2,15 +2,42 @@ import $ from "jquery";
+        $elem: JQuery;
+        callback: (data: string) => void;
+        message: Message;
+        extra_data: WidgetOptions["extra_data"];
```
```
extra_data: PollWidgetExtraData | ZFormExtraData | null
``` 
would be more readable.

LaPulgaaa replied:
We can live with this too if the core maintainers agree.

LaPulgaaa replied:
Sounds fair. !!

LaPulgaaa reviewed (commented):
@joydeep2001  This mostly lgtm .Just a few small optional improvement can be made.

joydeep2001 commented:
Thanks @LaPulgaaa for your review. I'm out of station. I'll be working on Monday.

LaPulgaaa commented:
@zulipbot add "integration review"

LaPulgaaa commented:
@timabbott LGTM; Ready for your final review.

timabbott commented:
I'm not convinced all the details are perfect, but generally this lgtm.
